graphix/channels.py

```diff
@@ -2,78 +2,137 @@
 
 from __future__ import annotations
 
+import copy
+import itertools
+import typing
+from typing import TYPE_CHECKING, SupportsIndex, TypeVar
+
 import numpy as np
+import numpy.typing as npt
 
-from graphix.linalg_validations import check_data_dims, check_data_normalization, check_data_values_type, check_rank
+from graphix import linalg_validations as lv
 from graphix.ops import Ops
 
+if TYPE_CHECKING:
+    from collections.abc import Iterable
+
+
+def _ilog2(n: int) -> int:
+    if n < 0:
+        raise ValueError("n must be non-negative.")
+    for p in itertools.count():
+        if n <= 2**p:
+            return p
+    raise RuntimeError("Unreachable.")
+
+
+# MEMO: Too much?
+_T = TypeVar("_T", bound=np.generic)
+
+
+class KrausData:
```

Comment:
I prefer `astype(copy=True)` here because `fronzen=True` does not save us from accidentally modifying `np.array` with multiple references.
As this class is exposed to users, we may be extra careful here.

Reply:
I think that, as a read-only data structure, we do not expect passed values to be copied.

Reply:
I don't think so as `dataclass` itself does not makes us feel that it's read-only. 
Actually `frozen=False` is invisible to users unless they wrongly try to reassign to it.

Reply:
I decided to use your idea here for simplicity.